akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/ActorContext.scala

```diff
@@ -154,4 +157,52 @@ trait ActorContext[T] { this: akka.actor.typed.javadsl.ActorContext[T] ⇒
    */
   def spawnAdapter[U](f: U ⇒ T): ActorRef[U]
 
+  /**
+   * Perform a single request-response message interaction with another actor, and transform the messages back to
+   * the protocol of this actor.
+   *
+   * The interaction has a timeout (to avoid resource a resource leak). If the timeout hits without any response it
+   * will be transformed to a message for this actor through the `failToOwnProtocol` function (this is the only
+   * "normal" way a `Failure` is passed to the function).
+   *
+   * For other messaging patterns with other actors, see [[spawnAdapter]].
+   *
+   * @param createMessage A function that creates a message for the other actor, containing the provided `ActorRef[Res]` that
+   *                      the other actor can send a message back through.
+   * @param responseToOwnProtocol Transforms the response from the `otherActor` into a message this actor understands.
+   *                              Should be a pure function but is executed inside the actor when the response arrives
+   *                              so can safely touch the actor internals. If this function throws an exception it is
+   *                              just as if the normal message receiving logic would throw.
+   *
+   * @tparam Req The request protocol, what the other actor accepts
+   * @tparam Res The response protocol, what the other actor sends back
+   */
+  def ask[Req, Res](
+    otherActor:    ActorRef[Req],
+    createMessage: ActorRef[Res] ⇒ Req
+  )(responseToOwnProtocol: Try[Res] ⇒ T)(implicit responseTimeout: Timeout, classTag: ClassTag[Res]): Unit
+
+  /**
+   * Implicit decoration of actorrefs to provide syntax sugar for [[ask]]
+   * @see [[Ask#ask]]
+   */
+  implicit final class Ask[Req](otherActor: ActorRef[Req]) {
+    /**
+     * Provides syntax sugar to call [[ask]] directly on an `ActorRef`:
+     * ```
+     * import ctx.Ask
```

Comment:
Does not seem like there’s a need to have it inside ctx? Does this not make the code clunky since I’m every behavior you will need to import the ask again and again? Could it be external, similar to how in untyped? To enable it in “this file”? Let me know if I missed dissuasion about it though..

Reply:
(Totally +1 for having it though, somewhere)

Reply:
If it's not inside `ctx` you'd have to take an `ActorContext` as an implicit parameter and cannot before hand know the type of the actor/behavior in the current scope, meaning that you have to mark the entire param list as implicit and possibly be more explicit with type parameters which seemed more clunky to me.

Maybe I'm missing some clever way to do this though.

Reply:
I see... will try to figure something out

Reply:
Yeah you're right... it's not possible to make it safe from completely the outside since we need the context..

One alternative which may be interesting is:

```
pingPong.ask(ctx, Ping) {
          case Success(pong) ⇒ Pong(ctx.self.path.name + "2", Thread.currentThread().getName)
          case Failure(ex)   ⇒ throw ex
        }```
```

that would not need any imports, but does not passing through the ctx. Personally this looks fine to me, but I wonder that people think. The implicit class wouldbe on ActorRef then,

Reply:
I thought about that alternative as well but in the end I like the

```
ctx.ask
```

more because you can read it as "Hey my context, ask that other actor", which seems to make sense.

Reply:
Perhaps we should just remove the implicits for now. Can always be added later.

I don't think the `import ctx.Ask` improves readability if you would have to add that line in front of each ask.